cirq/ops/projectors.py

```diff
@@ -0,0 +1,103 @@
+from typing import Any, Dict, Iterable, List, Tuple, TYPE_CHECKING, Union
+
+import numpy as np
+
+from cirq import value
+from cirq.ops import raw_types
+
+if TYPE_CHECKING:
+    import cirq
+    from cirq import protocols
+
+
+@value.value_equality
+class Projector(raw_types.Gate):
+    """A non-unitary gate that projects onto one or more qubits.
+
+    The input is a matrix representing the basis of the space we project onto.
+    The basis vectors need not be orthogonal, but they must be independent (i.e.
+    the matrix has full rank).
+
+    For example, if you want to project on |0>, you would provide the basis
+    [[1, 0]]. To project onto |10>, you would provide the basis [[0, 0, 1, 0]].
+    If you want to project on the space spanned by |10> and |11>, you could
+    provide the basis [[0, 0, 1, 0], [0, 0, 0, 1]].
+    """
+
+    def __init__(self,
+                 projection_basis: Union[List[List[float]], np.ndarray],
+                 qid_shape: Tuple[int, ...] = (2,),
+                 enforce_orthonormal_basis: bool = False):
+        """
+        Args:
+            projection_basis: a (p, 2**num_qubits) matrix that lists the
+                projection vectors, where p is the dimension of the subspace
+                we're projecting on. If you project onto a single vector, then
+                p = 1 and thus the matrix reduces to a single row.
+            qid_shape: Specifies the dimension of the projection. The default is
+                2 (qubit).
+            enforce_orthonormal_basis: Whether to enfore the input basis to be
+                orthogonal.
+
+        Raises:
+            ValueError: If the basis vector is empty.
+        """
+        projection_array = np.asarray(projection_basis)
+
+        if len(projection_array.shape) != 2:
+            raise ValueError('The input projection_basis must be a 2D array')
+
+        if enforce_orthonormal_basis:
+            B = projection_array @ projection_array.T.conj()
```

Comment:
You apply conjugation here and there is no testing of it - only real valued basis vectors are tested

Reply:
Do we really need the conjugation here?

Reply:
We expect (i, j) entry of B to be the inner product of the i-th and j-th row of `projection_array`. Since elements of `projection_array` can be complex(*) we need conjugation, because without it the elements of B aren't an inner product(**).

---
(*) For example the user may want to specify a projector onto one of the eigenvectors of Pauli Y operator, e.g. `projection_array=np.array([[1, 1j]])/np.sqrt(2)`.
(**) One of the requirements on the inner product is that <a, a> = 0 if and only if a = 0, but without conjugation <[1, 1j], [1, 1j]> = 0.

Reply:
Agreed - thanks for the explanation @viathor, I realized it right at the same time as well what's happening here :) 
However we do want to test it so I'm going to keep this thread open.

Reply:
I added two entries in the test: one that would (incorrectly) fail if the conjugation were missing and another that would (incorrectly) succeed if the conjugation were missing.

Given your remarks, I also took the liberty of adding a test in a different location. Maybe we want to make sure that even if we don't check for orthogonality, the computation of the projector is also working OK with complex number (I didn't have a test case, but now I have one).